libgnucash/engine/Account.cpp

```diff
@@ -5417,120 +5419,155 @@ gnc_account_imap_delete_account (GncImportMatchMap *imap,
-  product_difference is (1-a) * (1-b)
+/* The transaction to account matching algorithm uses bayesian inference to
+ * classify a sequence of tokens (observations), i.e. to find the account which
+ * most probably fits to the observed token sequence. See for instance
+ * https://en.wikipedia.org/wiki/Bayesian_inference.
+ *
+ * The underlying probalistic model is a token generator, which generates
+ * a sequence of identically distributed tokens of arbitrary length with
+ * probability P(T|A), which is the conditional probability of token T given
+ * account A. But the tokens are not independent random variables and therefore
+ * not i.i.d. Instead there is a strong relation between tokens from the same
+ * transaction. That means the joint probability of a token sequence
+ * P(T1,T2,...|A), which belong to the same transaction is not simply the
+ * product of the probabilities of the individual tokens
+ * P(T1|A)*P(T2|A)*... .
+ *
+ * The currently implemented algorithm handles each token individually and
+ * does not consider the joint probabilities between tokens. These statistics
+ * are not captured at the moment, only the token frequencies per account.
+ *
+ * The account classification of a token sequence is realized by an estimator,
+ * which maximizes the posterior probability P(A|T) of each token regarding
+ * the unknown account. This is known as maximum a posteriori (MAP) estimation.
+ *
+ * Since each token is handled individually, the estimation results for each
+ * token of the sequence have to be combined. This is done by calculating an
+ * average posterior probability mean(P(A|T1),P(A,T2),...). This final averaged
+ * probability is used for classification.
+ *
+ * This approach is enhanced by a token selection algorithm. This algorithm
+ * determines, which tokens should be used for calculating the averaged account
+ * probabilities and which tokens should be excluded.
  */
-struct AccountProbability
-{
-    double product; /* product of probabilities */
-    double product_difference; /* product of (1-probabilities) */
-};
 
-struct AccountTokenCount
+struct AccountTokenProbability
 {
     std::string account_guid;
     int64_t token_count; /** occurrences of a given token for this account_guid */
+    double probability; /* conditional probability of this account_guid for a given token */
 };
 
 /** total_count and the token_count for a given account let us calculate the
  * probability of a given account with any single token
  */
 struct TokenAccountsInfo
 {
-    std::vector<AccountTokenCount> accounts;
+    std::vector<AccountTokenProbability> accounts;
     int64_t total_count;
 };
 
-/** holds an account guid and its corresponding integer probability
-  the integer probability is some factor of 10
- */
-struct AccountInfo
-{
-    std::string account_guid;
-    int32_t probability;
-};
-
 static void
 build_token_info(char const * suffix, KvpValue * value, TokenAccountsInfo & tokenInfo)
```

Comment:
Should `KvpValue*` be `KvpValue const*`?

Reply:
I checked out your branch and using `KvpValue const *` compiles.

Reply:
West const, please, and no space between the type and `*`.
@bhardwajs You shouldn't need to compile to test that `const KvpValue*` (actually `const KvpValue* const` is appropriate since the function changes neither what `value` points to nor what it contains): `KvpValue::get()` is declared `const noexcept`.

Reply:
Changed as suggested